### ISSUE TITLE
Lint idea: `non_topologically_sorted_functions`

### ISSUE THREAD

**Issue by smoelius** [trailofbits/dylint#1582]
This would be a restriction lint that enforces a certain relative order among functions defined within a module.

First, the lint would ensure that "callers" appear before "callees". So, for example, the lint would warn about the following code because the definition of `foo` should appear before the definition of `bar`:
```rs
fn bar() { ... }

fn foo() { // `foo` should appear before `bar`
    bar();
}
```

Second, the lint would enforce relative order among callees. More specifically, in the definition of a function `foo`, if the first call to a function `bar` appears before the first call to a function `baz`, then the definition of `bar` should appear before the definition of `baz`. E.g., the lint would warn about the following code:
```rs
fn foo() {
    bar();
    baz();
}

fn baz() { ... }

fn bar() { ... } // `bar` should appear before `baz`
```

If there are multiple callers with calls to `bar` and `baz`, the order of the calls in the earlier defined caller would be preferred. E.g., in the following code, the order of the calls in `foo` would take precedence over the order of the calls in `qux`, and thus a warning would still be produced:
```rs
fn foo() {
    bar();
    baz();
}

fn qux() {
    baz();
    bar();
}

fn baz() { ... }

fn bar() { ... } // `bar` should still appear before `baz`
```

The lint would ignore calls to functions defined outside the module.

The lint should be `allow`able on a module-level basis.

**Comment by J4CKVVH173** [comment]
Hi, does it still need to be implemented? 
I can try to do

**Comment by smoelius** [comment]
Thanks!

It does still need to be implemented.

You can ignore this part, though, if you want:
> If there are multiple callers with calls to bar and baz, the order of the calls in the earlier defined caller would be preferred. E.g., in the following code, the order of the calls in foo would take precedence over the order of the calls in qux, and thus a warning would still be produced:

**Comment by smoelius** [comment]
Implemented in #1799
